lib/src/fty_shm.cc

```diff
@@ -71,65 +70,59 @@ int fty_get_polling_interval()
 static const char* shm_dir     = DEFAULT_SHM_DIR;
 static size_t      shm_dir_len = strlen(DEFAULT_SHM_DIR);
 
+// creates file name 
 static int prepare_filename(
-    char* buf, const char* asset, size_t a_len, const char* metric, size_t m_len, const char* type)
+    char* buf, size_t bufSize, const char* asset, const char* metric, const char* type)
 {
-    if (m_len + SEPARATOR_LEN + a_len > NAME_MAX) {
+
+    size_t assetLen = strlen(asset);
+    size_t metricLen = strlen(metric);
+    if (metricLen + SEPARATOR_LEN + assetLen > NAME_MAX) {
         errno = ENAMETOOLONG;
         return -1;
     }
-    if (memchr(asset, '/', a_len) || memchr(asset, SEPARATOR, a_len) || memchr(metric, '/', m_len) ||
-        memchr(metric, SEPARATOR, m_len)) {
+
+    if( (!asset || !(*asset)) || (!metric || !(*metric)) || (!type || !(*type)) )
+        return -1;
+
+    if (memchr(asset, '/', assetLen) || memchr(asset, SEPARATOR, assetLen) || memchr(metric, '/', metricLen) ||
+        memchr(metric, SEPARATOR, metricLen)) {
         errno = EINVAL;
         return -1;
     }
-    char* p = buf;
-    memcpy(p, shm_dir, shm_dir_len);
-    p += shm_dir_len;
-
-    *p++ = '/';
-    memcpy(p, type, strlen(type));
-    p += strlen(type);
-
-    *p++ = '/';
-    memcpy(p, metric, m_len);
-    p += m_len;
-    *p++ = SEPARATOR;
-    memcpy(p, asset, a_len);
-    p += a_len;
-    *p++ = '\0';
+
+    if(snprintf(buf, bufSize, "%s/%s/%s@%s",shm_dir, type, metric, asset) < 0)
+        return -1;
+    
     return 0;
 }
 
-// Write ttl and value to filename
-static int write_value(const char* filename, const char* value, const char* unit, int ttl)
+int fty_shm_write_metric(const char* asset, const char* metric, const char* value, const char* unit, int ttl)
 {
-    FILE* file = fopen(filename, "w");
-    if (file == nullptr)
-        return -1;
-    if (ttl < 0)
+    fty_proto_t * prt = fty_proto_new(FTY_PROTO_METRIC);
```

Comment:
what if fty_proto_new returns NULL?

Reply:
i'll add a null check here but it's a bigger problem if zmalloc returns null from fty_proto_new.

Reply:
we want to improve fty-shm, not criticize fty-proto.
Testing null ptr protect against segfault, and will improve quality
In general, checking a function returned value have the same effect, improve quality
it's a good habit to get into...